stacks-signer/src/v0/signer.rs

```diff
@@ -267,50 +268,49 @@ impl Signer {
                 .ok();
         }
 
-        let Some(sortition_state) = sortition_state else {
+        // If we have sortition state, run some additional checks
+        if let Some(sortition_state) = sortition_state {
+            match sortition_state.check_proposal(
+                stacks_client,
+                &self.signer_db,
+                &block_proposal.block,
+                miner_pubkey,
+            ) {
+                // Error validating block
+                Err(e) => warn!(
+                    "{self}: Error checking block proposal: {e:?}";
+                    "signer_sighash" => %signer_signature_hash,
+                    "block_id" => %block_proposal.block.block_id(),
+                ),
+                // Block proposal is bad
+                Ok(false) => {
+                    warn!(
+                        "{self}: Block proposal invalid";
+                        "signer_sighash" => %signer_signature_hash,
+                        "block_id" => %block_proposal.block.block_id(),
+                    );
+                    block_info.valid = Some(false);
+                }
+                // Block proposal passed check, still don't know if valid
+                Ok(true) => {}
+            }
+        } else {
             warn!(
                 "{self}: Cannot validate block, no sortition view";
                 "signer_sighash" => %signer_signature_hash,
                 "block_id" => %block_proposal.block.block_id(),
-            );
-            return;
+            )
```

Comment:
I think this should set `block_info.valid = Some(false)` -- if there's no sortition in the stacks-node, then the signer should just reject the block.

Reply:
If I add this, then the `signer::v0::block_proposal_rejection` fails with the error message above. I assumed the way to reject a block was to insert into `SignerDB` with `valid = Some(false)`, but maybe that's wrong?

Reply:
Ah, I think that the signer only considers sending back the response once `submit_block_for_validate` gets a validation event response.

In the check you added later on for `valid.is_none()` you should add an else branch which sends a response:

```
            debug!("{self}: Broadcasting a block response to stacks node: {block_response:?}");
            if let Err(e) = self
                .stackerdb
                .send_message_with_retry::<SignerMessage>(block_response.into())
            {
                warn!("{self}: Failed to send block rejection to stacker-db: {e:?}",);
            }
```

Reply:
Still getting a timeout on that test even after I add this. Here is the rejection message I'm sending:

```console
WARN [1719596013.862832] [stacks-signer/src/v0/signer.rs:321] [signer_runloop] Cycle #11 Signer #3: Broadcasting a block response to stacks node: Rejected(BlockRejection { reason: "Validation failed: InvalidBlock", reason_code: ValidationFailed(InvalidBlock), signer_signature_hash: ced4ed4e61eef9b2859ae1053bc5df414e7b3a901c927107ea7c66d1a81a7f28 })
```

I'll push the latest version with this change in a minute

Reply:
Maybe I have to send a `BlockValidateResponse` instead of a `BlockResponse`? Will try that now

Reply:
The `BlockValidateResponse` is the message going from the stacks-node to the signer. The `BlockResponse` should be the one going from the signer back to the miner.

Reply:
@obycode yeah I noticed that and deleted the comment. I check the resulting `StackerDBChunkAckData` from `send_message_with_retry()` and accepted is set to `true`. So stacker-db is getting and acknowledging the `BlockResponse` message, yet I'm still getting

```console
Timed out while waiting for block proposal response event
```

I think the issue is not that the miner is timing our waiting for a `BlockResponse`, but that the signer is timing out waiting for a `BlockValidateResponse`. If I enable the backtrace, it fails in this function:

```rust
fn wait_for_block_validate_response(&mut self, timeout: Duration) -> BlockValidateResponse {
    // ...
}
```